README
This is a "plugin" for the Video Disk Recorder (VDR).

Written by : Andreas Brachold <vdr07 AT deltab de>

Project's homepage:  http://projects.vdr-developer.org/wiki/plg-targavfd

This plugin for VDR is free software: you can redistribute it and/or 
modify it under the terms of the GNU General Public License as published 
by the Free Software Foundation, version 3 of the License.

This plugin for VDR is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY 
or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License 
for more details.

You should have received a copy of the GNU General Public License
along with this plugin for VDR. If not, see <http://www.gnu.org/licenses/>.

See the file COPYING for license information.

Description:
------------
targavfd is a plugin for the Video Disc Recorder and shows information
about the current state of VDR on Targa USB Graphic Vacuum Fluorescent Display.


Install libusb-1.0:
-------------------------------------
You need installed library libusb-1.0 to use the vdr-plugin-targavfd.
This library libusb-1.0 is used to access and interact with a USB device.

1. http://www.libusb.org/wiki/libusb-1.0

Install vdr-plugin-targavfd:
---------------------------
Unpack the vdr-targavfd tarball to $VDRSOURCE/PLUGINS/src directory.
Variable $VDRSOURCE stands for a directory, where source archive of VDR package 
was be decompressed.

#> cd $VDRSOURCE/PLUGINS/src
#> tar -xzvf vdr-targavfd-0.0.1.tgz
#> ln -s targavfd-0.0.1 targavfd

Compile the plugin (from VDR's source directory)

Change the working directory to VDR source root and compile this plugin 
like all other VDR plugins.

#> cd $VDRSOURCE
#> make plugins

After everything has been compiled without errors, copy the plugin to their 
target directory.

#> make install


Check follow requirements 
-------------------------
* Device still detected and writable

  #> lsusb -vd 19c2:6a11 | grep iProduct
  iProduct                2 Targa USB Graphic Vacuum Fluorescent Display

* Add a udev permission rule, if required (writable for group vdr) e.g.
  /etc/udev/rules.d/92-vfd.rules
  ACTION=="add", BUS=="usb", SYSFS{idVendor}=="19c2", SYSFS{idProduct}=="6a11", GROUP="vdr"

Start VDR with the plugin.
---------------------------
   Examples:
     vdr -P'targavfd'

Setup options
-------------
This options are editable at run time. (Menu: Setup > Plugin > targavfd ... )

* Brightness 
  - Set brightness from display. (Default: Half brightness)

* Default font 
  - Used font, there should installed like other FreeType supported fonts 
* Height of big font
* Height of small font

* Render mode
  - Single line
  - Dual lines
  - Single line (Channel names only)

* Exit mode
  - Do nothing - Just leave the "last" message there
  - Showing clock - Show the big clock
  - Turning display off - Blank the device completely
  - Show next timer - Show time and title of next timer
  - Show only present next timer - If present, show time and title of next timer 

* Show volume bargraph
  - Never - Show the volume bar never
  - Timed - Show the volume bar only a short time, if volume changed
  - Ever  - Show the volume bar every time

* Suspend display at night
  - Allow turn display off at night, depends selected mode and time frame.

Plugin SVDRP commands
---------------------
* HELP - List known commands
* OFF - Suspend driver of display.
* ON  - Resume driver of display.
* ICON [name] [on|off|auto] - Force state of icon. 

Use this commands like follow samples 
    #> svdrpsend.pl PLUG targavfd OFF
    220 vdr SVDRP VideoDiskRecorder ....
    250 driver suspended
    221 vdr closing connection

    #> svdrpsend.pl PLUG targavfd HELP ICON
    214-ICON [name] [on|off|auto]
    214-    Force state of icon. Names of icons are:
    214-    Play,Pause,Record,Message,Message,Mute...

This commands can have follow responses 
OFF :   250 driver suspended           
        251 driver already suspended
ON :    250 driver resumed
        251 driver already resumed
        554 driver could not resumed
ICON :  250 icon state 'auto'
        251 icon state 'on'
        252 icon state 'off'
*       501 unknown command

Spectrum analyzer visualization
-------------------------------
This plugin can show a spectrum analyzer visualization on audio playback.
As middleware you need the 'Sp'ectrum 'An'alyzer Plugin.

See also http://lcr.vdr-developer.org/htmls/span-plugin.html

Contribution of fonts
---------------------
See contrib/README



HISTORY
VDR Plugin 'targavfd' Revision History
-------------------------------------

2021-04-24: Version 0.3.2
- Add patch to support freetype 2.9.1 or higher
- Update plugin interface for vdr 2.3.3
- Drop legacy code prior vdr 1.7.3

2017-01-29: Version 0.3.1
- Add exception for nightly suspend for icons (Request #2459)
- Update plugin interface for vdr 2.3.2 (Request #2463)

2014-06-14: Version 0.3.0
- New render mode - multiple pages
- New option, show replay progress as bargraph (Request #666)
- Dual-Line-Mode: Eclipse long channel names (Request: #1815)
- Add a set of sample fonts (Request: #1821)
- Fix display wrong time for HD recording if they greater than one hour (Bug #1568)

2013-04-12: Version 0.2.1
- On amd64 linker binds not libusb (Bug #1338)
- Fix warning on amd64: format '%d' expects argument of type 'int'... [-Wformat]

2013-03-23: Version 0.2.0
- Adapt Makefile for new VDR >=1.7.36
- Fix bug: close crash without open device.
- Update plugin interface for vdr 1.7.26
- Closing of libusb corrected

2011-05-22: Version 0.1.0
- Drop dependency to libhid, now is libusb-1.0 used
- Wrong time on playback of HD recordings (Bug #517)
- Ignore empty subtitle on replay (Feature #484)

2010-11-02: Version 0.0.7
- Update italian translation (Support #445)

2010-10-31: Version 0.0.6
- Support spectrum analyzer visualization (need span-plugin)
- Allow turn display off at night (Feature #175)
- Add render mode "show channel names only" (Feature #426)
- Show next timer on lines (Feature #427)
- Update italian translation (Support #357)

2010-07-17: Version 0.0.5
- Add exit mode 'Show only present next timer' (Request #356)
- Add setup option to control volume bargraph
- Allow partial data transfer
- Select brightness by in plain language
- Update running clock frequently
- Don't show icons, if showing clock on exit

2010-04-28: Version 0.0.4
- Allow setup font size
- Show current time, on dual line mode

2010-04-18: Version 0.0.3
- Initial revision.



setup.c
/*
 * targavfd plugin for VDR (C++)
 *
 * (C) 2010-2017 Andreas Brachold <vdr07 AT deltab de>
 *
 * This targavfd plugin is free software: you can redistribute it and/or 
 * modify it under the terms of the GNU General Public License as published 
 * by the Free Software Foundation, version 3 of the License.
 *
 * See the files README and COPYING for details.
 *
 */

#include <string.h>
#include <stdio.h>
#include <stdlib.h>

#include "watch.h"
#include "setup.h"

#include <vdr/plugin.h>
#include <vdr/tools.h>
#include <vdr/config.h>
#include <vdr/i18n.h>

#define DEFAULT_ON_EXIT      eOnExitMode_BLANKSCREEN  /**< Blank the device completely */
#define DEFAULT_BRIGHTNESS   1
#define DEFAULT_WIDTH        96
#define DEFAULT_HEIGHT       16
#define DEFAULT_FONT         "Sans:Bold"
#define DEFAULT_TWO_LINE_MODE  eRenderMode_SingleLine
#define DEFAULT_BIG_FONT_HEIGHT   14
#define DEFAULT_SMALL_FONT_HEIGHT 7
#define DEFAULT_VOLUME_MODE   eVolumeMode_ShowEver    /**< Show the volume bar ever */
#define DEFAULT_SUSPEND_MODE   eSuspendMode_Never      /**< Suspend display never */

/// The one and only Stored setup data
cVFDSetup theSetup;

/// Ctor, load default values
cVFDSetup::cVFDSetup(void)
:  m_cWidth(DEFAULT_WIDTH)
,  m_cHeight(DEFAULT_HEIGHT)
{
  m_nOnExit = DEFAULT_ON_EXIT;
  m_nBrightness = DEFAULT_BRIGHTNESS;
  m_nRenderMode = DEFAULT_TWO_LINE_MODE;
  m_nBigFontHeight = DEFAULT_BIG_FONT_HEIGHT;
  m_nSmallFontHeight = DEFAULT_SMALL_FONT_HEIGHT;
  m_nVolumeMode = DEFAULT_VOLUME_MODE;
  m_nSuspendMode = DEFAULT_SUSPEND_MODE;
  m_nSuspendTimeOn = 2200;
  m_nSuspendTimeOff = 800;
  m_bSuspend_Timed = 1;   /**< Suspend display, resume short time */
  m_bSuspend_Icons = 1;   /**< Suspend icons */

  strncpy(m_szFont,DEFAULT_FONT,sizeof(m_szFont));
}

cVFDSetup::cVFDSetup(const cVFDSetup& x)
:  m_cWidth(DEFAULT_WIDTH)
,  m_cHeight(DEFAULT_HEIGHT)
{
  *this = x;
}

cVFDSetup& cVFDSetup::operator = (const cVFDSetup& x)
{
  m_nOnExit = x.m_nOnExit;
  m_nBrightness = x.m_nBrightness;

  m_nRenderMode = x.m_nRenderMode;
  m_nBigFontHeight = x.m_nBigFontHeight;
  m_nSmallFontHeight = x.m_nSmallFontHeight;

  m_nVolumeMode = x.m_nVolumeMode;
  m_nSuspendMode = x.m_nSuspendMode;
  m_nSuspendTimeOn = x.m_nSuspendTimeOn;
  m_nSuspendTimeOff = x.m_nSuspendTimeOff;

  m_bSuspend_Timed = x.m_bSuspend_Timed;
  m_bSuspend_Icons = x.m_bSuspend_Icons;

  strncpy(m_szFont,x.m_szFont,sizeof(m_szFont));

  return *this;
}

bool cVFDSetup::SetupParseInt(const char *szName, const char *szValue,
    const char *szKey,
    const int nMin, const int nMax, const int nDefault,
    int& nValue) 
{
  if(!strcasecmp(szName, szKey)) {
    int n = atoi(szValue);
    if ((n < nMin) || (n >= nMax)) {
        esyslog("targaVFD: %s must be between %d and %d, using default %d",
               szKey, nMin, nMax-1, nDefault);
        n = nDefault;
    }
    nValue = n;
    dsyslog("targaVFD: %s set to %d", szKey, nValue);
    return true;
  }
  return false;
}

/// compare profilenames and load there value
bool cVFDSetup::SetupParse(const char *szName, const char *szValue)
{
  if(SetupParseInt(szName, szValue, "OnExit", eOnExitMode_SHOWMSG, eOnExitMode_LASTITEM, DEFAULT_ON_EXIT, m_nOnExit)) { return true; }
  if(SetupParseInt(szName, szValue, "Brightness", 0, 2, DEFAULT_BRIGHTNESS, m_nBrightness)) { return true; }

  // Font
  if(!strcasecmp(szName, "Font")) {
    if(szValue) {
      cStringList fontNames;
      cFont::GetAvailableFontNames(&fontNames);
      if(fontNames.Find(szValue)>=0) {
        strncpy(m_szFont,szValue,sizeof(m_szFont));
        dsyslog("targaVFD: %s set to %s", szName, m_szFont);
        return true;
      }
    }
    esyslog("targaVFD:  %s '%s' not found, using default %s",
        szName, szValue, DEFAULT_FONT);
    strncpy(m_szFont,DEFAULT_FONT,sizeof(m_szFont));
    return true;
  }

  if(SetupParseInt(szName, szValue, "BigFont", 5, 24, DEFAULT_BIG_FONT_HEIGHT, m_nBigFontHeight)) { return true; }
  if(SetupParseInt(szName, szValue, "SmallFont", 5, 24, DEFAULT_SMALL_FONT_HEIGHT, m_nSmallFontHeight)) { return true; }
  if(SetupParseInt(szName, szValue, "TwoLineMode", eRenderMode_SingleLine, eRenderMode_LASTITEM, DEFAULT_TWO_LINE_MODE, m_nRenderMode)) { return true; }
  if(SetupParseInt(szName, szValue, "VolumeMode", eVolumeMode_ShowNever, eVolumeMode_LASTITEM, DEFAULT_VOLUME_MODE, m_nVolumeMode)) { return true; }
  if(SetupParseInt(szName, szValue, "SuspendMode", eSuspendMode_Never, eSuspendMode_LASTITEM, DEFAULT_SUSPEND_MODE, m_nSuspendMode)) { return true; }
  if(SetupParseInt(szName, szValue, "SuspendTimed", 0, 2, 1, m_bSuspend_Timed)) { return true; }
  if(SetupParseInt(szName, szValue, "SuspendIcons", 0, 2, 1, m_bSuspend_Icons)) { return true; }
  if(SetupParseInt(szName, szValue, "SuspendTimeOn", 0, 2400, 2200, m_nSuspendTimeOn)) { return true; }
  if(SetupParseInt(szName, szValue, "SuspendTimeOff", 0, 2400, 800, m_nSuspendTimeOff)) { return true; }

  //Unknow parameter
  return false;
}


// --- cVFDMenuSetup --------------------------------------------------------
void cVFDMenuSetup::Store(void)
{
  theSetup = m_tmpSetup;

  SetupStore("OnExit",     theSetup.m_nOnExit);
  SetupStore("Brightness", theSetup.m_nBrightness);
  SetupStore("Font",       theSetup.m_szFont);
  SetupStore("BigFont", theSetup.m_nBigFontHeight);
  SetupStore("SmallFont", theSetup.m_nSmallFontHeight);
  SetupStore("TwoLineMode",theSetup.m_nRenderMode);
  SetupStore("VolumeMode", theSetup.m_nVolumeMode);
  SetupStore("SuspendMode", theSetup.m_nSuspendMode);
  SetupStore("SuspendTimed", theSetup.m_bSuspend_Timed);
  SetupStore("SuspendIcons", theSetup.m_bSuspend_Icons);
  SetupStore("SuspendTimeOn", theSetup.m_nSuspendTimeOn);
  SetupStore("SuspendTimeOff", theSetup.m_nSuspendTimeOff);
}

cVFDMenuSetup::cVFDMenuSetup(cVFDWatch* pDev)
: m_tmpSetup(theSetup)
, m_pDev(pDev)
{
  SetSection(tr("targavfd"));

  cFont::GetAvailableFontNames(&fontNames);
  fontNames.Insert(strdup(DEFAULT_FONT));
  fontIndex = max(0, fontNames.Find(m_tmpSetup.m_szFont));

  static const char * szBrightness[3];
  szBrightness[0] = tr("Show nothing");
  szBrightness[1] = tr("Half brightness");
  szBrightness[2] = tr("Full brightness");

  Add(new cMenuEditStraItem (tr("Brightness"),
        &m_tmpSetup.m_nBrightness,
        memberof(szBrightness), szBrightness));

  static const char * szRenderMode[eRenderMode_LASTITEM];
  szRenderMode[eRenderMode_SingleLine] = tr("Single line");
  szRenderMode[eRenderMode_DualLine] = tr("Dual lines");
  szRenderMode[eRenderMode_SingleTopic] = tr("Only topic");
  szRenderMode[eRenderMode_MultiPage] = tr("Multiple pages");

  Add(new cMenuEditStraItem(tr("Render mode"),
        &m_tmpSetup.m_nRenderMode,    
        memberof(szRenderMode), szRenderMode));

  Add(new cMenuEditStraItem(tr("Default font"),
        &fontIndex, fontNames.Size(), &fontNames[0]));
  Add(new cMenuEditIntItem (tr("Height of big font"),
        &m_tmpSetup.m_nBigFontHeight,        
        5, 24));
  Add(new cMenuEditIntItem (tr("Height of small font"),
        &m_tmpSetup.m_nSmallFontHeight,        
        5, 24));

  static const char * szSuspendMode[eSuspendMode_LASTITEM];
  szSuspendMode[eSuspendMode_Never] = tr("Never");
  szSuspendMode[eSuspendMode_Ever]  = tr("Only per time");

  Add(new cMenuEditStraItem (tr("Suspend display at night"),
        &m_tmpSetup.m_nSuspendMode,
        memberof(szSuspendMode), szSuspendMode));
  Add(new cMenuEditBoolItem(tr("Resume on user activities"),
        &m_tmpSetup.m_bSuspend_Timed,
        tr("No"), tr("Yes")));
  Add(new cMenuEditBoolItem(tr("Hide icons"),
        &m_tmpSetup.m_bSuspend_Icons,
        tr("No"), tr("Yes")));
  Add(new cMenuEditTimeItem (tr("Beginning of suspend"),
        &m_tmpSetup.m_nSuspendTimeOn));
  Add(new cMenuEditTimeItem (tr("End time of suspend"),
        &m_tmpSetup.m_nSuspendTimeOff));

  static const char * szVolumeMode[eVolumeMode_LASTITEM];
  szVolumeMode[eVolumeMode_ShowNever] = tr("Never");
  szVolumeMode[eVolumeMode_ShowTimed] = tr("as volume short time");
  szVolumeMode[eVolumeMode_ShowEver]  = tr("as volume");
  szVolumeMode[eVolumeMode_Progress]  = tr("as replay progress");

  Add(new cMenuEditStraItem (tr("Show bargraph"),
        &m_tmpSetup.m_nVolumeMode,
        memberof(szVolumeMode), szVolumeMode));

  static const char * szExitModes[eOnExitMode_LASTITEM];
  szExitModes[eOnExitMode_SHOWMSG]      = tr("Do nothing");
  szExitModes[eOnExitMode_SHOWCLOCK]    = tr("Showing clock");
  szExitModes[eOnExitMode_BLANKSCREEN]  = tr("Turning display off");
  szExitModes[eOnExitMode_NEXTTIMER]    = tr("Show next timer");
  szExitModes[eOnExitMode_NEXTTIMER_BLANKSCR] = tr("Show only present next timer");

  Add(new cMenuEditStraItem (tr("Exit mode"),
        &m_tmpSetup.m_nOnExit,
        memberof(szExitModes), szExitModes));
}

eOSState cVFDMenuSetup::ProcessKey(eKeys nKey)
{
  if(nKey == kOk) {
    // Store edited Values
    Utf8Strn0Cpy(m_tmpSetup.m_szFont, fontNames[fontIndex], sizeof(m_tmpSetup.m_szFont));
    if (0 != strcmp(m_tmpSetup.m_szFont, theSetup.m_szFont)
        || m_tmpSetup.m_nRenderMode != theSetup.m_nRenderMode
        || ( m_tmpSetup.m_nRenderMode != eRenderMode_DualLine && (m_tmpSetup.m_nBigFontHeight != theSetup.m_nBigFontHeight))
        || ( m_tmpSetup.m_nRenderMode == eRenderMode_DualLine && (m_tmpSetup.m_nSmallFontHeight != theSetup.m_nSmallFontHeight))
      ) {
        m_pDev->SetFont(m_tmpSetup.m_szFont, 
                        m_tmpSetup.m_nRenderMode == eRenderMode_DualLine ? true : false, 
                        m_tmpSetup.m_nBigFontHeight, 
                        m_tmpSetup.m_nSmallFontHeight);
    }
  }
  return cMenuSetupPage::ProcessKey(nKey);
}


